src/gil.rs

```diff
@@ -494,6 +494,31 @@ impl EnsureGIL {
     }
 }
 
+/// Value with concurrent access protected by the GIL.
+///
+/// This is a synchronization primitive based on Python's global interpreter lock (GIL).
+/// It ensures that only one thread at a time can access the inner value via shared references.
+/// It can be combined with interior mutability to obtain mutable references.
+pub struct GILProtected<T> {
+    value: T,
+}
+
+impl<T> GILProtected<T> {
+    /// Place the given value under the protection of the GIL.
+    pub const fn new(value: T) -> Self {
+        Self { value }
+    }
+
+    /// Gain access to the inner value by giving proof of having acquired the GIL.
+    pub fn with_gil<'py>(&'py self, _py: Python<'py>) -> &'py T {
+        &self.value
+    }
+}
+
+unsafe impl<T> Send for GILProtected<T> where T: Send {}
+
+unsafe impl<T> Sync for GILProtected<T> where T: Send {}
```

Comment:
I'm wondering if `T: Send` requirement can be dropped here? As I see it, `GILProtected<T>` offers an API which always is thread-safe, `T` is irrelevant.

Reply:
If I clone an `Rc<()>` on thread A, put it inside a `GILProtected<T>` and send `&GILProtected<Rc<()>>` to thread B, then on thread B, I can do `.with_gil(py).clone()` to produce an `Rc<()>` thereby racily increment the reference count against thread A with never had the GIL in the first place?

I have to admit to have a hard time phrasing this more generally, but I have the same problem with the equivalent bound for `Mutex`. But just because of the counter example if think the bound is required.

Reply:
Another example would GUI objects bound to the main thread where even calling methods through shared references on them on other threads would not work, independently of whether those calls are synchronized, i.e. exclusively interleave, using the GIL.

Reply:
Right, understood. Those counter-examples are clear enough to me, thanks.

Reply:
I think @birkenfeld had some fitting general terms for this: https://users.rust-lang.org/t/solved-how-to-move-non-send-between-threads-or-an-alternative/19928/9

Reply:
Yes, this bound is required. If `GILProtected` is `Sync` then `&GILProtected` can be sent to another thread, and sending `&GILProtected<Rc<T>>` to another thread is unsound.